### PR TITLE
feat: add arc bundle and arc publish commands (F-006)

### LINKED ISSUE #61: feat(A-500): arc bundle + arc publish (Phase 1 bootstrap)
## Context

Completing the end-to-end package lifecycle. F-4 (install from registry) is done — arc can download, verify, and install from metafactory. But there's no way to **publish** packages. This is the publisher-side convergence.

**Phase 1 (this issue):** Bootstrap API path — `arc bundle` creates tarballs, `arc publish` uploads to R2 and registers versions via existing `POST /api/v1/` endpoints with bearer token auth.

**Phase 2 (future):** Migrate to DD-97 seam contract with Sigstore keyless OIDC, `/intake/v1/` endpoints, and presigned R2 uploads.

## Spec

Full spec: `.specify/specs/f-6-bundle-and-publish/spec.md`

### `arc bundle`
- [ ] Read arc-manifest.yaml from package directory
- [ ] Validate manifest meets publishing requirements (name, version, type)
- [ ] Create .tar.gz excluding .git, node_modules, .env, .specify, test/
- [ ] Compute and display SHA-256
- [ ] Display bundle summary (file count, size, type)
- [ ] `--output` flag for custom tarball path
- [ ] Warn if no README.md, size > 10MB, or missing description

### `arc publish`
- [ ] Require authentication (bearer token from `arc login`)
- [ ] Bundle the package (reuse `arc bundle` logic)
- [ ] Upload tarball to `POST /api/v1/storage/upload`
- [ ] Verify client SHA-256 matches server SHA-256
- [ ] Register version via `POST /api/v1/packages/@scope/name/versions`
- [ ] Auto-create package if it does not exist yet
- [ ] Include README.md in version payload (render-on-publish, CP-7)
- [ ] Handle 409 Conflict for immutable versions (DD-14)
- [ ] `--dry-run` flag: validate and show what would be published
- [ ] `--tarball` flag: publish from existing bundle
- [ ] `--source` flag: target specific metafactory source
- [ ] Clean up temp tarball on all code paths

### Tests
- [ ] Unit: tarball creation, exclusion patterns, manifest validation
- [ ] Unit: upload with mocked storage endpoint
- [ ] Unit: version registration with mocked API
- [ ] Command: `arc bundle` CLI integration
- [ ] Command: `arc publish` CLI integration (mocked API)
- [ ] Error paths: no manifest, no auth, version exists, SHA mismatch

## Gate

`arc publish` from a directory with arc-manifest.yaml uploads a tarball to the metafactory registry, and `arc install @scope/name` on another machine downloads and installs it. Full round-trip.

## Authority

- `.specify/specs/f-6-bundle-and-publish/spec.md` (full spec)
- metafactory `POST /api/v1/storage/upload` returns `{ sha256, r2_key, size_bytes }` (verified in storage.ts:101-108)
- metafactory `POST /api/v1/packages/:scope/:name/versions` requires `{ version, sha256, r2_key, manifest, size_bytes, readme? }` (verified in packages.ts:647-668)
- DD-14 (content-addressed immutable storage)
- DD-46 (three-tier access: publish requires IDENTIFIED+)

## Related

- F-4 (install from registry) — the consumer side, already done
- #59 (seam contract) — Phase 2 migration target
- #52 (anonymous install) — install side can proceed independently

### PR DESCRIPTION
**Pull request by jcfischer**
## Summary

- Implements `arc bundle` and `arc publish` commands — publisher-side complement to F-4 (install from registry)
- Phase 1 bootstrap: bearer token auth, direct upload through Worker, content-addressed R2 storage
- Spec: `.specify/specs/f-6-bundle-and-publish/spec.md` (reviewed by Luna, two rounds)

### New commands

| Command | Description |
|---------|-------------|
| `arc bundle [path]` | Create `.tar.gz` from package directory with exclusions |
| `arc bundle --output <path>` | Custom output path for tarball |
| `arc publish [path]` | Bundle → upload → register version on metafactory |
| `arc publish --dry-run` | Validate without uploading |
| `arc publish --tarball <path>` | Publish from existing tarball |
| `arc publish --scope <ns>` | Override publish namespace |
| `arc publish --source <name>` | Target specific metafactory source |

### Key features

- **Default exclusions**: `.git`, `node_modules`, `.env`, `.DS_Store`, `test/`, `dist/`, etc.
- **Manifest overrides**: `bundle.exclude` and `bundle.include` in arc-manifest.yaml
- **SHA-256 verification**: client-computed hash verified against server response
- **Size limits**: 50MB hard limit (client pre-check), 10MB warning
- **Auto-create package**: first publish auto-creates the package entry
- **Scope resolution**: `--scope` flag > manifest `namespace` > `/auth/me` API
- **Idempotent uploads**: 409 on duplicate content treated as success (DD-14)
- **Retry on server error**: registration auto-retries once on 5xx

### New files

| File | Purpose |
|------|---------|
| `src/lib/bundle.ts` | Tarball creation, exclusions, checksum, validation |
| `src/lib/publish.ts` | Upload, scope resolution, package creation, version registration |
| `src/commands/bundle.ts` | `arc bundle` command handler |
| `src/commands/publish.ts` | `arc publish` command handler |
| `test/unit/bundle.test.ts` | 22 unit tests for bundle library |
| `test/unit/publish.test.ts` | 18 unit tests for publish library (mocked HTTP) |
| `test/commands/bundle.test.ts` | 8 command tests for arc bundle |
| `test/commands/publish.test.ts` | 9 command tests for arc publish (mocked API) |

## Test plan

- [x] 57 new tests pass (unit + command)
- [x] 457 total tests pass (0 regressions)
- [ ] Manual E2E: `arc bundle` on a real package → inspect tarball contents
- [ ] Manual E2E: `arc publish --dry-run` → verify validation output
- [ ] Manual E2E: `arc publish` against staging metafactory → `arc install` on clean machine

Closes #61

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lib/bundle.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  return {
+    success: true,
+    tarballPath: tarballName,
+    sha256,
+    sizeBytes,
+    fileCount,
+    manifest,
+    warnings,
+  };
+}
```
Comment (mellanon) [inline src/lib/bundle.ts:188]: **[warning/quality]** `createBundle()` is ~120 lines with 8+ distinct responsibilities: read manifest, validate, check README, build tar args, create tarball, compute checksum, get size, check size limits. Each numbered step is a natural decomposition point — extract at least `buildTarArgs()` and the size-check/cleanup logic.

Guideline: functions over 50 lines need decomposition (CodeQuality §4), and functions over 25 lines are flagged (Architecture §1).

File: src/lib/publish.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+/** Resolve the publish scope (namespace) with three-tier priority */
+export async function resolvePublishScope(
+  manifest: ArcManifest,
+  source: RegistrySource,
+  cliScope?: string,
+): Promise<string | null> {
+  // 1. CLI --scope flag
+  if (cliScope) return cliScope;
+
```
Comment (mellanon) [inline src/lib/publish.ts:55]: **[warning/architecture]** The existing `metafactory-api.ts:106` has a shared `buildHeaders(source)` helper for constructing the Authorization header. This file builds `Authorization: Bearer ${source.token}` inline in three separate functions (`uploadBundle`, `ensurePackageExists`, `registerVersion`).

Reuse `buildHeaders` from metafactory-api, or extract a shared utility to avoid triple duplication of auth header construction.

File: src/lib/publish.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    if (!resp.ok) {
+      return { success: false, sha256: "", r2Key: "", sizeBytes: 0, error: body.error ?? `Upload failed: HTTP ${resp.status}` };
+    }
+
+    // Verify SHA-256 match
+    if (body.sha256 !== clientSha256) {
+      return {
+        success: false,
```
Comment (mellanon) [inline src/lib/publish.ts]: **[warning/architecture]** Timeout values `10_000` and `120_000` are inline magic numbers. The existing `metafactory-api.ts` uses a named constant `REQUEST_TIMEOUT_MS = 10_000`. Follow the same pattern — define `UPLOAD_TIMEOUT_MS = 120_000` and `API_TIMEOUT_MS = 10_000` (or reuse the existing constant).

File: src/lib/publish.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+      signal: AbortSignal.timeout(10_000),
+    });
+
+    if (!resp.ok) return null;
+
+    const body = (await resp.json()) as { namespace?: string };
+    return body.namespace ?? null;
+  } catch {
+    return null;
+  }
```
Comment (mellanon) [inline src/lib/publish.ts:75]: **[suggestion/quality]** This catch block silently discards the error from the `/auth/me` API call. When scope resolution fails, users get "Could not determine publish scope" with no debugging context.

The existing `metafactory-api.ts` uses `debugLog()` for these cases. At minimum, log the error to stderr or via debugLog so users can diagnose API connectivity issues with `ARC_DEBUG=1`.

File: src/lib/publish.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+      return { success: false, sha256: "", r2Key: "", sizeBytes: 0, error: 'Not authenticated. Run "arc login" first.' };
+    }
+
+    if (resp.status === 413) {
+      return { success: false, sha256: "", r2Key: "", sizeBytes: 0, error: "Package too large for server." };
+    }
+
+    const body = (await resp.json()) as { sha256: string; r2_key: string; size_bytes: number; error?: string };
+
+    // 409 = content already exists — treat as success (idempotent)
```
Comment (mellanon) [inline src/lib/publish.ts:110]: **[suggestion/security]** API response bodies are type-asserted with `as` without runtime validation. For example, `body.sha256` and `body.r2_key` could be undefined if the server returns unexpected JSON, and these values flow into later logic (`registerVersion` payload).

The existing `metafactory-api.ts:200-203` validates `body.packages && Array.isArray(body.packages)` before use. Follow the same defensive pattern — at minimum check that `body.sha256` and `body.r2_key` are truthy strings before treating the upload as successful.

File: src/lib/publish.ts [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    const createBody = await createResp.json().catch(() => ({})) as { error?: string };
+    if (createResp.status === 403) {
+      return { exists: false, created: false, error: `You do not own namespace @${scope}. Complete identity verification at meta-factory.ai.` };
+    }
+
+    return { exists: false, created: false, error: createBody.error ?? `Failed to create package: HTTP ${createResp.status}` };
+  } catch {
+    return { exists: false, created: false, error: "Network error checking package existence." };
```
Comment (mellanon) [inline src/lib/publish.ts]: **[suggestion/quality]** Both catch blocks in the retry loop (lines 186 and 155) discard the actual error, returning generic "Network error" messages. Including `(err as Error).message` would significantly aid debugging — especially for `registerVersion` which is the most failure-prone step (happens after upload, so data is partially committed).

Pattern from existing code: `metafactory-api.ts:216` names the caught error `_err` and uses `debugLog()`.

**Comment by mellanon** [comment]
## Code Review: the-metafactory/arc#62

**SOP: pr-review | PR: the-metafactory/arc#62 | Lenses: CodeQuality, Security, Architecture**

### Cross-Cutting Findings

**[warning/quality] Dead field: `PublishValidation.scope`**
`src/types.ts` declares `scope?: string` on `PublishValidation`, but `validateForPublish()` never populates it. Either remove it or set it.

**[warning/duplication] README existence checked in two places with different code**
`src/lib/bundle.ts:155-160` inlines a README existence check (`existsSync` for three variants). `src/lib/publish.ts:14` defines `README_VARIANTS` and `extractReadme()` which searches the same three variants. The bundle code should call `extractReadme()` (and check the return for null) or at least import the shared `README_VARIANTS` constant.

**[warning/duplication] Three near-identical `createPackage` test helpers**
- `test/commands/bundle.test.ts:22` — `createPackage()`
- `test/commands/publish.test.ts:38` — `createPackage()`
- `test/unit/bundle.test.ts:23` — `createMockPackage()`

All create a temp dir with a manifest YAML and optional README. Extract to `test/helpers/test-env.ts` alongside the existing `createTestEnv()`, `createMockSkillRepo()`, etc.

**[nit/duplication] `mockFetch` duplicated across test files**
`test/commands/publish.test.ts:9`, `test/unit/publish.test.ts:42`, and `test/unit/metafactory-api.test.ts:116` all have slightly different `mockFetch` implementations. Consider extracting a shared `test/helpers/mock-fetch.ts`.

**[nit/quality] Dead test setup in `test/commands/publish.test.ts:138`**
Line 138 calls `saveSources(env.paths.sourcesPath, metafactorySource(undefined as any))` and then immediately overwrites it on line 140 with a proper tokenless source. The first call is dead code — remove it.

### What's Good

- Clean command/lib separation — commands handle CLI, lib handles business logic, types are shared
- Comprehensive test coverage: 57 tests covering happy paths, error states, edge cases (409 idempotency, 5xx retry, scope resolution priority)
- SHA-256 verification on upload is properly implemented
- Proper cleanup of temp tarballs via `finally` blocks
- Good error messages with actionable guidance ("Run arc login first", "bump the version")
- Scope resolution three-tier priority is well-designed
- Types are well-structured with clear success/error result patterns

### Findings Summary

| Severity | Count | Lens |
|----------|-------|------|
| critical | 0 | — |
| warning | 5 | quality (2), architecture (2), duplication (1) |
| suggestion | 3 | quality (2), security (1) |
| nit | 2 | duplication (1), quality (1) |

### Verdict: Request Changes

10 findings total — mostly pattern consistency issues (reuse existing `buildHeaders`, `debugLog`, timeout constants) and test helper duplication. No security vulnerabilities or critical defects. Solid implementation that needs a polish pass to align with existing codebase patterns.

**Review by mellanon — CHANGES_REQUESTED** [review]
10 findings across CodeQuality, Security, and Architecture lenses. Key themes: reuse existing patterns (buildHeaders, debugLog, timeout constants), decompose long functions, extract duplicated test helpers, add runtime validation on API responses. No critical defects — solid implementation needing a polish pass.

**Comment by jcfischer** [comment]
## Luna Review — All 10 Findings Addressed

Pushed to `feat/F-006-bundle-publish` (commit 7c1c02e). 457 tests pass, 0 regressions.

**Warnings (5) — fixed:**
- `createBundle()` decomposed: extracted `buildTarArgs()` and `getBundleStats()` helpers
- `buildHeaders` duplication: added `buildPublishHeaders()` in publish.ts, used in all 3 functions
- Magic timeouts: added `API_TIMEOUT_MS = 10_000` and `UPLOAD_TIMEOUT_MS = 120_000` constants
- Dead `PublishValidation.scope` field: removed from types.ts
- README check duplication: `README_VARIANTS` exported from bundle.ts, publish.ts imports it

**Suggestions (3) — fixed:**
- `resolvePublishScope` catch: now logs `debugLog(\`/auth/me scope resolution failed: ${err.message}\`)`
- Upload response validation: both 409 and 200 paths check `!body.sha256 || !body.r2_key` before use
- Catch blocks: both `uploadBundle` and `registerVersion` include `(err as Error).message`

**Nits (2) — fixed:**
- Dead test setup (`saveSources(undefined as any)`): removed from publish.test.ts
- `mockFetch` duplication: extracted to `test/helpers/mock-fetch.ts`, 2 test files updated

**Bonus:** `createPackageDir` extracted to `test/helpers/test-env.ts` — all 3 duplicate test helpers consolidated.

**Review by mellanon — APPROVED** [review]
## Re-review: All 10 findings resolved ✅

**Lenses applied:** CodeQuality + Security + Architecture

### Finding verification (all 10 addressed)

| # | Original finding | Resolution |
|---|-----------------|------------|
| 1 | `createBundle()` 120+ lines | `buildTarArgs()` + `getBundleStats()` extracted |
| 2 | Duplicated `buildHeaders` across 3 functions | `buildPublishHeaders()` consolidated |
| 3 | Magic timeout numbers | `API_TIMEOUT_MS` + `UPLOAD_TIMEOUT_MS` named constants |
| 4 | Silent catch in scope resolution | `debugLog()` with `(err as Error).message` |
| 5 | API response bodies unvalidated | `!body.sha256 \|\| !body.r2_key` runtime guards |
| 6 | Dead `PublishValidation.scope` field | Removed from type |
| 7 | README check duplication | `README_VARIANTS` exported + imported |
| 8 | 3× `createPackageDir` helpers | Consolidated in `test/helpers/test-env.ts` |
| 9 | 2× `mockFetch` implementations | Consolidated in `test/helpers/mock-fetch.ts` |
| 10 | Dead `saveSources(undefined as any)` | Removed |

### Fix commit quality

The fix commit is surgical — each change targets exactly one finding. No unrelated modifications, no regressions. Extracted helpers are clean and well-named. Test helper consolidation is properly structured with shared imports.

**Zero new findings.** Ready to merge.
